### DISCUSSION TITLE
Cross-compiling with libuv for Windows on Linux with mingw?

### DISCUSSION THREAD

**Discussion by T-X** [libuv/libuv#4482] [Q&A] [answered]
I'm on Debian Sid and trying to create my first portable library with the help of libuv and autotools. Generally I seem to be able to compile an executable for Windows here on Debian thanks to mingw. I'm using this to build for instance: `autoreconf -i && ./configure --host=x86_64-w64-mingw32 && make`.

However while I do have libuv1t64 and libuv1-dev installed and can build for Debian/Linux just fine with them, I'm now missing libuv for the Windows/mingw build. What are recommended ways for cross compiling for Windows on Debian/Linux with libuv? Any way I can install libuv for mingw here on Debian? Or would I need to use a more heavy approach, like wine+cygwin and then compile in there?

**Comment by bnoordhuis** [comment]
I'm using pretty much the same stanza as the one you wrote and that works fine for me so...

> I'm now missing libuv for the Windows/mingw build

...I'm not sure what this means, exactly. What files are missing? Where would you expect them to turn up?

**Reply by T-X** [reply]
Hi @bnoordhuis, thanks for the reply! The library I'm working on is this one: https://github.com/T-X/libsap/tree/wip-libuv (still work in progress, not quite cleaned up yet :-) )

These things build and run fine on Debian Sid, as expected:
1) `./configure --with-uv && make clean && make`
2) `./configure --without-uv && make clean && make`

However `./configure (--with-uv) --host=x86_64-w64-mingw32 && make clean && make` does not:
```
$ ./configure --with-uv --host=x86_64-w64-mingw32 && make clean && make                                <─(Tue,Jul30)─┘
checking for x86_64-w64-mingw32-gcc... x86_64-w64-mingw32-gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether x86_64-w64-mingw32-gcc accepts -g... yes
checking for x86_64-w64-mingw32-gcc option to enable C11 features... none needed
checking whether x86_64-w64-mingw32-gcc understands -c and -o together... yes
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for zlib.h... yes
checking for deflate in -lz... yes
checking for json-c/json.h... no
checking for blake2.h... no
checking for uv.h... no
...
```

It won't find a uv.h and then will obviously fail with building my epoll fallback with mingw, as expected.

I think I would expect to have uv.h and a libuv.dll in /usr/x86_64-w64-mingw32/include/ and /usr/x86_64-w64-mingw32/lib/ on Debian. But I don't see a Debian package for libuv-mingw-w64 and libuv-mingw-w64-dev in apt either.

@bnoordhuis do you have something like this? Or how would you recommend building an app depending on libuv for Windows on Linux / Debian Sid?

**Reply by T-X** [reply]
And overall the selection of libraries for mingw on Debian Sid seems rather limited:

```
$ apt-cache search mingw | grep "lib.*ming.*-dev"
libassuan-mingw-w64-dev - IPC library for the GnuPG components -- Windows port
libgcrypt-mingw-w64-dev - LGPL Crypto library - Windows development
libgpg-error-mingw-w64-dev - library of error values and messages in GnuPG (Windows development)
libksba-mingw-w64-dev - X.509 and CMS support library (Windows development)
libz-mingw-w64-dev - compression library - Windows development files
libnpth-mingw-w64-dev - replacement for GNU Pth using system threads (Windows dev)
```

I guess for a beginner like myself it would be very helpful to have a libuv-mingw-w64-dev in there, too.

(But again, if there are other, suggested, maybe slightly longer ways to get there, I'd be all ears.)

**Reply by bnoordhuis** [reply, marked as answer]
Ah okay, so your issue is basically that libuv-dev isn't installed? I can't really help you with that, we don't package libuv for distros.

I just vendor libuv with my projects. Then, when I'm cross-compiling, it builds (and builds against) the in-tree copy.

**Reply by T-X** [reply]
Thanks for the reply again! I think I found a solution for me for now. Luckily, libuv is pretty light regarding its dependencies, so I basically got and built libuv from Git for mingw locally, like:

```
$ git clone https://github.com/libuv/libuv.git
$ cd libuv
$ ./autogen.sh && ./configure --host=x86_64-w64-mingw32 && make
$ cd ../libsap
$ autoreconf -i && ./configure --host=x86_64-w64-mingw32 'CFLAGS=-I../libuv/include -L../libuv/.libs/' && make
```

Which was easier than I had expected :D. I also wrote an email to a Debian maintainer who currently maintains a couple of mingw packages. Let's see if that leads to anything. It seems that Arch Linux is a little further in that regard, they have packaged some more libraries for mingw, including libuv.
